src/project.coffee

```diff
@@ -39,6 +39,20 @@ class Project extends Model
     @emitter = new Emitter
     @buffers ?= []
 
+    # Mapping from the real path of a {Directory} to a {Promise} that resolves
+    # to either a {Repository} or null. Ideally, the {Directory} would be used
+    # as the key; however, there can be multiple {Directory} objects created for
+    # the same real path, so it is not a good key.
```

Comment:
> I'm not sure if Project should be maintaining that map or if a DirectoryProvider should be responsible for its own cache.

You're right. It should definitely be the provider. But I'm comfortable assuming the provider does this correctly and making that part of the contract.

I think the name should clarify the key type. So if we're changing this to be indexed by path, then it should probably be `repositoryPromisesByPath` so as not to mislead. It would be great to actually make it a `Map` and fix the aliasing problems in pathwatcher :soon:.